src/vg_conn.erl

```diff
@@ -152,33 +153,28 @@ handle_request(?FETCH_REQUEST, Role, <<_ReplicaId:32/signed-integer, _MaxWaitTim
                CorrelationId, Socket) when Role =:= tail orelse Role =:= solo ->
     {[{Topic, [{Partition, Offset, MaxBytes} | _]} | _], _} = vg_protocol:decode_array(fun decode_topic/1, Rest),
     {SegmentId, Position} = vg_log_segments:find_segment_offset(Topic, Partition, Offset),
-
-    lager:info("at=fetch_request topic=~s partition=~p offset=~p segment_id=~p position=~p",
-              [Topic, Partition, Offset, SegmentId, Position]),
-
-    File = vg_utils:log_file(Topic, Partition, SegmentId),
-    {ok, Fd} = file:open(File, [read, binary, raw]),
-    try
-        SendBytes = filelib:file_size(File) - Position,
-        Bytes =
-            case MaxBytes of
-                0 -> SendBytes;
-                _ -> min(SendBytes, MaxBytes)
-            end,
-        ErrorCode = 0,
-        HighWaterMark = vg_topics:lookup_hwm(Topic, Partition),
-        Response = vg_protocol:encode_fetch_response(Topic, Partition, ErrorCode, HighWaterMark, Bytes),
-        lager:info("sending hwm=~p bytes=~p", [HighWaterMark, Bytes]),
-        gen_tcp:send(Socket, [<<(Bytes + 4 + iolist_size(Response)):32/signed-integer,
-                                CorrelationId:32/signed-integer>>, Response]),
-        {ok, _} = file:sendfile(Fd, Socket, Position, Bytes, [])
-        %% maybe catch any errors here and report them?
-    after
-        file:close(Fd)
-    end;
-handle_request(?FETCH_REQUEST, _ , <<_ReplicaId:32/signed, _MaxWaitTime:32/signed,
+    do_fetch(Topic, Partition, Offset, SegmentId, Position, unlimited,
+             CorrelationId, Socket, MaxBytes);
+handle_request(?FETCH2_REQUEST, Role, <<_ReplicaId:32/signed-integer, _MaxWaitTime:32/signed-integer,
+                                       _MinBytes:32/signed-integer, Rest/binary>>,
+               CorrelationId, Socket) when Role =:= tail orelse Role =:= solo ->
+    {[{Topic, [{Partition, Offset, MaxBytes, MaxIndex} | _]} | _], _} = vg_protocol:decode_array(fun decode_topic2/1, Rest),
+    {SegmentId, Position} = vg_log_segments:find_segment_offset(Topic, Partition, Offset),
+    {EndSegmentId, EndPosition} = vg_log_segments:find_segment_offset(Topic, Partition, MaxIndex + 1),
+    Fetch =
+        case SegmentId of
+            %% max on this segment, limit fetch
+            EndSegmentId ->
+                {limited, EndPosition - Position};
+            %% some higher segment, unlimited fetch
+            _ ->
+                unlimited
+        end,
+    do_fetch(Topic, Partition, Offset, SegmentId, Position, Fetch,
+             CorrelationId, Socket, MaxBytes);
+handle_request(Fetch, _ , <<_ReplicaId:32/signed, _MaxWaitTime:32/signed,
                                      _MinBytes:32/signed, Rest/binary>>,
-               CorrelationId, Socket) ->
+               CorrelationId, Socket) when Fetch == ?FETCH_REQUEST orelse Fetch == ?FETCH2_REQUEST ->
```

Comment:
Why does would we need to match on both here? Shouldn't it just be `FETCH_REQUEST`?